### PR TITLE
feat(js): openai tool and function call support on streaming

### LINKED ISSUE #114: [JS] Tool Calls from Streaming

### PR DESCRIPTION
**Pull request by mikeldking**
resolves #114 

Adds support for function and tool calls invoked during streaming for openai

### CODE REVIEW

File: js/.eslintrc.js [hunk 1]
```diff
@@ -29,6 +29,7 @@ module.exports = {
         varsIgnorePattern: "^_",
       },
     ], // ignore unused variables starting with underscore
-    eqeqeq: ["error", "always"],
+    eqeqeq: ["error", "smart"],
```
Comment (mikeldking) [inline js/.eslintrc.js:32]: is allows != for things like nulls which makes sense

File: js/examples/openai/src/index.ts [hunk 2]
```diff
@@ -12,13 +12,15 @@ const openai = new OpenAI({
...
+    // eslint-disable-next-line no-console
     console.log(choice.message);
   });
   chatCompletion = await openai.chat.completions.create({
     messages: [{ role: "user", content: "Tell me a joke" }],
     model: "gpt-3.5-turbo",
   });
   chatCompletion.choices.forEach((choice) => {
+    // eslint-disable-next-line no-console
     console.log(choice.message);
```
Comment (axiomofjoy) [inline js/examples/openai/src/index.ts:24]: Assuming we want to nix these?

**Review by RogerHYang — APPROVED** [review]

File: js/packages/openinference-instrumentation-openai/test/openai.test.ts [hunk 3]
```diff
@@ -421,34 +422,273 @@ describe("OpenAIInstrumentation", () => {
...
-  "llm.invocation_parameters": "{"model":"gpt-3.5-turbo","tools":[{"type":"function","function":{"name":"getCurrentLocation","parameters":{"type":"object","properties":{}},"description":"Get the current location of the user."}},{"type":"function","function":{"name":"getWeather","parameters":{"type":"object","properties":{"location":{"type":"string"}}},"description":"Get the weather for a location."}}],"tool_choice":"auto","stream":false}",
-  "llm.model_name": "gpt-3.5-turbo-0613",
-  "llm.output_messages.0.message.content": "The weather in Boston this week is expected to be rainy with a temperature of 52 degrees.",
+  "llm.invocation_parameters": "{"model":"gpt-3.5-turbo","tools":[{"type":"function","function":{"name":"getCurrentLocation","parameters":{"type":"object","properties":{}},"description":"Get the current location of the user."}},{"type":"function","function":{"name":"getWeather","description":"Get the weather for a location.","parameters":{"type":"object","properties":{"location":{"type":"string"}}}}}],"stream":true}",
+  "llm.model_name": "gpt-3.5-turbo",
+  "llm.output_messages.0.message.content": "",
   "llm.output_messages.0.message.role": "assistant",
```
Comment (axiomofjoy) [inline js/packages/openinference-instrumentation-openai/test/openai.test.ts:573]: Role is not correct here, there are specialized "tool_call" and "function_call" roles.

File: js/packages/openinference-instrumentation-openai/test/openai.test.ts [hunk 4]
```diff
@@ -421,34 +422,273 @@ describe("OpenAIInstrumentation", () => {
...
+        "llm.input_messages.0.message.content": "What's the weather today?",
+        "llm.input_messages.0.message.role": "user",
+        "llm.invocation_parameters": "{"model":"gpt-3.5-turbo","functions":[{"name":"getWeather","description":"Get the weather for a location.","parameters":{"type":"object","properties":{"location":{"type":"string"}}}},{"name":"getCurrentLocation","description":"Get the current location of the user.","parameters":{"type":"object","properties":{}}}],"stream":true}",
+        "llm.model_name": "gpt-3.5-turbo",
+        "llm.output_messages.0.message.content": "",
+        "llm.output_messages.0.message.function_call_arguments_json": "{}",
+        "llm.output_messages.0.message.function_call_name": "getWeather",
+        "llm.output_messages.0.message.role": "assistant",
```
Comment (axiomofjoy) [inline js/packages/openinference-instrumentation-openai/test/openai.test.ts:687]: Same as above.

File: js/examples/openai/src/index.ts (hunk 2 above)
Comment (mikeldking) [reply js/examples/openai/src/index.ts:24]: No, it's okay - It's just that no console is good for the instrumentations but not the examples. We will figure out an override

File: js/packages/openinference-instrumentation-openai/test/openai.test.ts (hunk 3 above)
Comment (mikeldking) [reply js/packages/openinference-instrumentation-openai/test/openai.test.ts:573]: This is not the tool call but the assistant saying invoke a tool

```
// {role: "user",      content: "How's the weather this week?"}
// {role: "assistant", tool_calls: [{type: "function", function: {name: "getCurrentLocation", arguments: "{}"}, id: "123"}
// {role: "tool",      name: "getCurrentLocation", content: "Boston", tool_call_id: "123"}
// {role: "assistant", tool_calls: [{type: "function", function: {name: "getWeather", arguments: '{"location": "Boston"}'}, id: "1234"}]}
// {role: "tool",      name: "getWeather", content: '{"temperature": "50degF", "preciptation": "high"}', tool_call_id: "1234"}
// {role: "assistant", content: "It's looking cold and rainy - you might want to wear a jacket!"}
//
// Final content: "It's looking cold and rainy - you might want to wear a jacket!"
```

File: js/packages/openinference-instrumentation-openai/test/openai.test.ts (hunk 4 above)
Comment (mikeldking) [reply js/packages/openinference-instrumentation-openai/test/openai.test.ts:687]: Yeah not the tool call, the first message is the assistant saying to make a tool / function call

**Pull request merged**